### PR TITLE
SOLR-15149: model creation errors fixes

### PR DESCRIPTION
**Pull request by alessandrobenedetti**
# Description

LTR integration gives obsvure and misleading error messages when creating models under certain circumstances

# Solution

better exception handling

# Tests

Tests to cover the exceptional use cases have been added

# Checklist

Please review the following and check all that apply:

- [X] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [X] I have created a Jira issue and added the issue ID to my pull request title.
- [X] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [X] I have developed this patch against the `master` branch.
- [X] I have run `./gradlew check`.
- [X] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/model/LTRScoringModel.java [hunk 1]
```diff
@@ -108,7 +108,7 @@ public static LTRScoringModel getInstance(SolrResourceLoader solrResourceLoader,
         SolrPluginUtils.invokeSetters(model, params.entrySet());
       }
     } catch (final Exception e) {
-      throw new ModelException("Model type does not exist " + className, e);
+      throw new ModelException("Model creation failed for " + className, e);
```
Comment (cpoerschke) [inline solr/contrib/ltr/src/java/org/apache/solr/ltr/model/LTRScoringModel.java:111]: minor: how about not saying "creation" but perhaps "loading" or something to that effect? technically the "creation" i.e. storage of the model's JSON succeeded (if i remember things right) but it's the "loading" or "using" of the model's JSON that encounters the error. so the user seeing the error needs to look at the content of what they uploaded but not at the mechanics of how the uploaded/created the model.

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/model/LinearModel.java [hunk 2]
```diff
@@ -80,10 +80,10 @@
 
   public void setWeights(Object weights) {
     @SuppressWarnings({"unchecked"})
-    final Map<String,Double> modelWeights = (Map<String,Double>) weights;
+    final Map<String,Number> modelWeights = (Map<String, Number>) weights;
     for (int ii = 0; ii < features.size(); ++ii) {
       final String key = features.get(ii).getName();
-      final Double val = modelWeights.get(key);
+      final Number val = modelWeights.get(key);
```
Comment (cpoerschke) [inline solr/contrib/ltr/src/java/org/apache/solr/ltr/model/LinearModel.java:86]: neat!

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/store/rest/ManagedModelStore.java [hunk 3]
```diff
@@ -294,11 +294,18 @@ private static void initWrapperModel(SolrResourceLoader solrResourceLoader,
...
-        : featureStore.get(featureName));
+  private static Feature lookupFeatureFromFeatureMap(Map<String, Object> featureMap, FeatureStore featureStore)
+  {
+    final String featureName = (String) featureMap.get(NAME_KEY);
+    Feature extractedFromStore = featureName == null ? null : featureStore.get(featureName);
+    if (extractedFromStore == null) {
+      if (featureStore.getFeatures().isEmpty()) {
+        throw new ModelException("Feature Store not found: " + featureStore.getName());
+      } else {
+        throw new ModelException("Feature:" + featureName + " not found in store: " + featureStore.getName());
```
Comment (cpoerschke) [inline solr/contrib/ltr/src/java/org/apache/solr/ltr/store/rest/ManagedModelStore.java:305]:
```suggestion
        throw new ModelException("Feature: " + featureName + " not found in store: " + featureStore.getName());
```

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/store/rest/ManagedModelStore.java [hunk 4]
```diff
@@ -294,11 +294,18 @@ private static void initWrapperModel(SolrResourceLoader solrResourceLoader,
...
-    final String featureName = (String)featureMap.get(NAME_KEY);
-    return (featureName == null ? null
-        : featureStore.get(featureName));
+  private static Feature lookupFeatureFromFeatureMap(Map<String, Object> featureMap, FeatureStore featureStore)
+  {
+    final String featureName = (String) featureMap.get(NAME_KEY);
+    Feature extractedFromStore = featureName == null ? null : featureStore.get(featureName);
+    if (extractedFromStore == null) {
+      if (featureStore.getFeatures().isEmpty()) {
+        throw new ModelException("Feature Store not found: " + featureStore.getName());
```
Comment (cpoerschke) [inline solr/contrib/ltr/src/java/org/apache/solr/ltr/store/rest/ManagedModelStore.java:303]:
```suggestion
        throw new ModelException("Missing or empty feature store: " + featureStore.getName());
```

File: solr/contrib/ltr/src/test-files/modelExamples/linear-model_notExistentFeature.json [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  ],
+  "params": {
+    "weights": {
+      "title": 0.0000000000,
+      "description": 0.1000000000,
+      "keywords": 0.2000000000,
+      "popularity": 0.3000000000,
+      "text": 0.4000000000,
+      "queryIntentPerson": 0.1231231,
+      "queryIntentCompany": 0.12121211
```
Comment (cpoerschke) [inline solr/contrib/ltr/src/test-files/modelExamples/linear-model_notExistentFeature.json:21]: how about using `"notExist1"` and `"notExist2"` here for the weights too?

File: solr/contrib/ltr/src/test-files/modelExamples/multipleadditivetreesmodel_unknownFeature.json [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    "features":[
+        { "name": "notExist1"},
+        { "name": "notExist2"}
+    ],
+    "params":{
+        "trees": [
+            {
+                "weight" : "1f",
+                "root": {
+                    "feature": "matchedTitle",
```
Comment (cpoerschke) [inline solr/contrib/ltr/src/test-files/modelExamples/multipleadditivetreesmodel_unknownFeature.json:13]:
```suggestion
                    "feature": "notExist1",
```

File: solr/contrib/ltr/src/test-files/modelExamples/multipleadditivetreesmodel_unknownFeature.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+                    "threshold": "0.5f",
+                    "left" : {
+                        "value" : "-100"
+                    },
+                    "right": {
+                        "feature" : "constantScoreToForceMultipleAdditiveTreesScoreAllDocs",
```
Comment (cpoerschke) [inline solr/contrib/ltr/src/test-files/modelExamples/multipleadditivetreesmodel_unknownFeature.json:19]:
```suggestion
                        "feature" : "notExist2",
```

**Review by cpoerschke — COMMENTED** [review]
Thanks for improving the error messages and test coverage!

File: solr/contrib/ltr/src/java/org/apache/solr/ltr/model/LTRScoringModel.java (hunk 1 above)
Comment (alessandrobenedetti) [reply solr/contrib/ltr/src/java/org/apache/solr/ltr/model/LTRScoringModel.java:111]: agreed!

**Comment by alessandrobenedetti** [comment]
Thanks @cpoerschke  for the review and contributions! The Jira linking is highly appreciated!
I update the Pull Request with your suggestions, as soon as I have a green build, I proceed to the merge!

**Pull request merged**